### DISCUSSION TITLE
React Storefront

### DISCUSSION THREAD

**Discussion by karolkielecki** [saleor/saleor#8736] [Show and tell]
If you didn't have the chance to explore our latest open source project - React Storefront - now is the time to catch up! The app focuses on best practices in e-commerce, including SSR & ISR, image optimization. It's fast and powered by Next.js!
Repo: https://github.com/saleor/react-storefront
Tutorial: https://learn.saleor.io/

And good luck with creating your own app!

--------
- [Twitter](https://twitter.com/getsaleor/status/1467826801439383565)

**Comment by webpioneer** [comment]
Hi @karolkielecki Which branch is used in https://reactstorefront.vercel.app/
When i clone the repo and use the main branch, i don't get the cart part working and get an Error on the cart page.

Thank you in advance.

**Reply by karolkielecki** [reply]
@webpioneer Hi! Please share more details on the error you receive and, ideally, some screenshots. More info is needed to reproduce it.

**Comment by slamer59** [comment]
I would be interested in seeing the comparison with **saleor-storefront**  ?

What is in Reactstorefront which is not in saleor-storefront ? What will be integrated ? Which plugins are already available ? Can we still use dashboard to manage (I think so but...) ? 

I am considering saleor-storefront which will be deprecated and reactstorefront seems to be in alpha. 

What about https://github.com/saleor/nextjs-commerce ? I am confused on where to look at. 

Disclaimer I am not an expert of React / Nextjs :)

**Reply by slamer59** [reply]
**nextjs-commerce** is a fork from a NextJS project. It seems to be an alternative / showcase as I read here https://saleor.io/blog/saleor-and-nextjs-how-the-two-are-a-powerhouse-combination-145/

**Reply by krzysztofwolski** [reply]
Hello!
In the headless approach, all clients (storefronts, dashboards) are just applications using the same API, but independently. This means you can swap or add as many storefronts as you want, e.g., completely different storefronts for Europe and the USA, different mobile apps, etc. You’ll still be able to manage them from the same dashboard or create an additional dedicated dashboard to handle your specific edge case. There are no “hidden APIs” or restrictions.
Some clarification:
- saleor/storefront - it’s the current storefront. Currently, it has the best feature coverage, but we are planning to sunset this project in favor of react-storefront.
- saleor/react-storefront - it’s a new project. Once we achieve feature parity with the old one, it will be our main storefront. I think it will happen around Q1/22.
- saleor/nextjs-commerce - [Nextjs commerce](https://nextjs.org/commerce) is an universal storefront for multiple backends maintained by Vercel.
Our main purpose for the saleor/react-storefront repo is to create a good starter pack for real storefronts.
We leverage our experience working with our clients and community. Many of them would like to be able to modify storefronts heavily and adapt them to new designs. Also, some of them would like to use different technologies (css/ui libraries are quite polarizing across the community ;) ). The old saleor/storefront has more visual flair, but it comes at the cost of more complicated modifications.
That’s why we decided to create a new project (react-storefront) from scratch:
- we want to provide a solid API commutation base (Apollo + Graphql Code Generator, Saleor SDK) and handle edge cases of the checkout management
- Leverage server-side rendering of the Nextjs (with the incremental rendering which will help to minimalize build time, and allow us better caching and SEO)
- completely refactored checkout. Now it’s “one view” and simplified logic for easier integrating with new payment providers. Right now we only support test and Stripe payments, we’ll be integrating with the Apple Pay/GPay and Adyen soon.
- A simple design. It’s easier to modify designs according to your needs. We used the Tailwind library and it’s pretty user-friendly. TBH, I really do recommend using it. Had my objections at first, but I was positively surprised! Also, it should be easy to swap it to another one of your team preferences.
As you mentioned, it’s still in alpha since we have many moving parts and would like to create an extensible structure. We’ve just implemented channels and now finalizing the translations. After that, we’ll be working on payments.

**Reply by slamer59** [reply]
Hello and Happy New Year ! 
Thanks for your exhaustive feedback it clarifies a lot my thought. I already learned a lot just looking at saleor architecture. Your advise are valuable. 

For this reason, I would try to answer my best when I can to the community.

**Comment by rcalvarez37201** [comment]
Hi
I need help to configure and allow PWA in my saleor project
Thanks in advance

**Reply by peelar** [reply]
> Hi I need help to configure and allow PWA in my saleor project Thanks in advance

Hello @rcalvarez37201 👋 

Please describe your problem and we will do our best trying to help.

**Comment by yetqdn** [comment]
Hello all,
I have a bug so could you tell me how to fix:
This is my terminal when i run pnpm dev:
root@vps:/home/saleor/react-storefront# pnpm dev

> saleor-storefront@0.0.0 dev /home/saleor/react-storefront
> env-cmd --no-override -f .env env-cmd --no-override -f .env.local --fallback pnpm run _dev


> saleor-storefront@0.0.0 _dev /home/saleor/react-storefront
> turbo run build --filter=@saleor/checkout-storefront --cache-dir=.turbo && turbo run dev --parallel --cache-dir=.turbo

• Packages in scope: @saleor/checkout-storefront
• Running build in 1 packages
@saleor/ui-kit:build: cache miss, executing 8e07329d84ecc721
checkout-common:build: cache miss, executing 27281c5a79d3590e
checkout-common:build:
checkout-common:build: > checkout-common@0.1.0 build /home/saleor/react-storefront/packages/checkout-common
checkout-common:build: > NODE_ENV=production rollup -c
checkout-common:build:
@saleor/ui-kit:build:
@saleor/ui-kit:build: > @saleor/ui-kit@0.0.0 build /home/saleor/react-storefront/packages/ui-kit
@saleor/ui-kit:build: > rollup -c
@saleor/ui-kit:build:
@saleor/ui-kit:build:
@saleor/ui-kit:build: src/index.ts → dist/cjs/index.js, dist/esm/index.js...
checkout-common:build:
checkout-common:build: src/index.ts → dist/esm/index.js, dist/cjs/index.js...
checkout-common:build: created dist/esm/index.js, dist/cjs/index.js in 11.9s
checkout-common:build:
checkout-common:build: ./dist/esm/index.d.ts → dist/index.d.ts...
checkout-common:build: created dist/index.d.ts in 111ms
@saleor/ui-kit:build: created dist/cjs/index.js, dist/esm/index.js in 25.3s
@saleor/ui-kit:build:
@saleor/ui-kit:build: ./dist/esm/index.d.ts → dist/index.d.ts...
@saleor/ui-kit:build: (!) Unresolved dependencies
@saleor/ui-kit:build: https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
@saleor/ui-kit:build: @lib/globalTypes (imported by dist/esm/components/Checkbox/Checkbox.d.ts, dist/esm/components/Chip/Chip.d.ts, dist/esm/components/IconButton/IconButton.d.ts, dist/esm/components/InfoTip/InfoTip.d.ts, dist/esm/components/Radio/Radio.d.ts, dist/esm/components/Select/Select.d.ts, dist/esm/components/Switch/Switch.d.ts)
@saleor/ui-kit:build: created dist/index.d.ts in 296ms
@saleor/checkout-storefront:build: cache miss, executing 31025787bac7e1ea
@saleor/checkout-storefront:build:
@saleor/checkout-storefront:build: > @saleor/checkout-storefront@0.1.0 build /home/saleor/react-storefront/packages/checkout-storefront
@saleor/checkout-storefront:build: > NODE_ENV=production rollup -c
@saleor/checkout-storefront:build:
@saleor/checkout-storefront:build:
@saleor/checkout-storefront:build: src/index.tsx → dist/esm/index.js...
@saleor/checkout-storefront:build: (!) Circular dependencies
@saleor/checkout-storefront:build: ../../node_modules/.pnpm/@react-aria+i18n@3.6.0_react@18.2.0/node_modules/@react-aria/i18n/dist/module.js -> ../../node_modules/.pnpm/@react-aria+i18n@3.6.0_react@18.2.0/node_modules/@react-aria/i18n/dist/useMessageFormatter.module.js -> ../../node_modules/.pnpm/@react-aria+i18n@3.6.0_react@18.2.0/node_modules/@react-aria/i18n/dist/module.js
@saleor/checkout-storefront:build: src/views/Checkout/index.tsx -> src/views/Checkout/Checkout.tsx -> src/views/index.tsx -> src/views/Checkout/index.tsx
@saleor/checkout-storefront:build: created dist/esm/index.js in 47.2s
@saleor/checkout-storefront:build:
@saleor/checkout-storefront:build: ./dist/esm/index.d.ts → dist/index.d.ts...
@saleor/checkout-storefront:build: (!) Unresolved dependencies
@saleor/checkout-storefront:build: https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
@saleor/checkout-storefront:build: @/checkout-storefront/providers/AppConfigProvider/types (imported by dist/esm/pages/Root.d.ts)
@saleor/checkout-storefront:build: created dist/index.d.ts in 61ms

 Tasks:    3 successful, 3 total
Cached:    0 cached, 3 total
  Time:    1m19.566s

• Packages in scope: , @saleor/checkout-storefront, @saleor/ui-kit, checkout, checkout-common, env-vars, eslint-config-checkout, saleor-app-checkout, storefront, tsconfig
• Running dev in 10 packages
saleor-app-checkout:dev: cache bypass, force executing 28fd3d8d96726a9d
checkout-common:dev: cache bypass, force executing d42a794760f51d2e
checkout:dev: cache bypass, force executing 5522bbe3f3d1d57a
storefront:dev: cache bypass, force executing 16676f88d7761727
@saleor/ui-kit:dev: cache bypass, force executing 328be879ba474157
@saleor/checkout-storefront:dev: cache bypass, force executing b6330851707157d5
saleor-app-checkout:dev:
saleor-app-checkout:dev: > saleor-app-checkout@0.0.0 dev /home/saleor/react-storefront/apps/saleor-app-checkout
saleor-app-checkout:dev: > NODE_OPTIONS='--inspect --require env-vars' next dev -p 3001
saleor-app-checkout:dev:
@saleor/ui-kit:dev:
@saleor/ui-kit:dev: > @saleor/ui-kit@0.0.0 dev /home/saleor/react-storefront/packages/ui-kit
@saleor/ui-kit:dev: > rollup -c -w
@saleor/ui-kit:dev:
saleor-app-checkout:dev: Debugger listening on ws://127.0.0.1:9229/c4519d22-3add-42a0-b379-dba344d5fb2c
saleor-app-checkout:dev: For help, see: https://nodejs.org/en/docs/inspector
checkout-common:dev:
checkout-common:dev: > checkout-common@0.1.0 dev /home/saleor/react-storefront/packages/checkout-common
checkout-common:dev: > rollup -c -w
checkout-common:dev:
saleor-app-checkout:dev: ready - started server on 0.0.0.0:3001, url: http://localhost:3001
saleor-app-checkout:dev: info  - Loaded env from /home/saleor/react-storefront/apps/saleor-app-checkout/.env.development
saleor-app-checkout:dev: info  - Loaded env from /home/saleor/react-storefront/apps/saleor-app-checkout/.env
storefront:dev:
storefront:dev: > storefront@0.1.0 dev /home/saleor/react-storefront/apps/storefront
storefront:dev: > next dev --port 3000
storefront:dev:
checkout:dev:
checkout:dev: > checkout@0.1.0 dev /home/saleor/react-storefront/apps/checkout
checkout:dev: > react-scripts start
checkout:dev:
@saleor/checkout-storefront:dev:
@saleor/checkout-storefront:dev: > @saleor/checkout-storefront@0.1.0 dev /home/saleor/react-storefront/packages/checkout-storefront
@saleor/checkout-storefront:dev: > concurrently --kill-others 'next dev -p 3002' 'rollup -w -c'
@saleor/checkout-storefront:dev:
storefront:dev: ready - started server on 0.0.0.0:3000, url: http://localhost:3000
storefront:dev: info  - Loaded env from /home/saleor/react-storefront/apps/storefront/.env
saleor-app-checkout:dev: warn  - You have enabled experimental feature (esmExternals) in next.config.js.
saleor-app-checkout:dev: warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.
saleor-app-checkout:dev:
@saleor/checkout-storefront:dev: [0] ready - started server on 0.0.0.0:3002, url: http://localhost:3002
@saleor/checkout-storefront:dev: [0] info  - Loaded env from /home/saleor/react-storefront/packages/checkout-storefront/.env
storefront:dev: info  - SWC minify release candidate enabled. https://nextjs.link/swcmin
checkout-common:dev: rollup v2.79.0
checkout-common:dev: bundles src/index.ts → dist/esm/index.js...
@saleor/checkout-storefront:dev: [0] warn  - Detected next.config.js, no exported configuration found. https://nextjs.org/docs/messages/empty-configuration
@saleor/ui-kit:dev: rollup v2.79.0
@saleor/ui-kit:dev: bundles src/index.ts → dist/cjs/index.js, dist/esm/index.js...
saleor-app-checkout:dev: info  - Disabled SWC as replacement for Babel because of custom Babel configuration ".babelrc" https://nextjs.org/docs/messages/swc-disabled
saleor-app-checkout:dev: info  - Using external babel configuration from /home/saleor/react-storefront/apps/saleor-app-checkout/.babelrc
@saleor/checkout-storefront:dev: [1] rollup v2.79.0
@saleor/checkout-storefront:dev: [1] bundles src/index.tsx → dist/esm/index.js...
checkout:dev: (node:259469) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
checkout:dev: (Use `node --trace-deprecation ...` to show where the warning was created)
checkout:dev: (node:259469) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
checkout:dev: Starting the development server...
checkout:dev:
@saleor/ui-kit:dev: Killed
@saleor/ui-kit:dev:  ELIFECYCLE  Command failed with exit code 137.
@saleor/ui-kit:dev: ERROR: command finished with error: command (/home/saleor/react-storefront/packages/ui-kit) pnpm run dev exited (1)
@saleor/checkout-storefront:dev: [0] next dev -p 3002 exited with code SIGINT
@saleor/checkout-storefront:dev: --> Sending SIGTERM to other processes..
@saleor/checkout-storefront:dev: [1] rollup -w -c exited with code SIGTERM
command (/home/saleor/react-storefront/packages/ui-kit) pnpm run dev exited (1)

 Tasks:    0 successful, 1 total
Cached:    0 cached, 1 total
  Time:    53.231s

 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed with exit code 1.
root@vps:/home/saleor/react-storefront#

Thanks

**Reply by peelar** [reply]
Hello @yetqdn 👋 

Can you please clone the latest version of the repository and check if the issue still occurs? I see that you are using some older version (your shell references `saleor-storefront@0.0.0`).

**Reply by yetqdn** [reply]
How can I clone the latest version. I just did the git clone command: git clone https://github.com/saleor/react-storefront.git

I make command: git branch and I see branch is canary, so What branch I must checkout

Thanks
